pipeline/coaddition.py

```diff
@@ -0,0 +1,423 @@
+
+from sep import Background
+
+from models.provenance import Provenance
+from models.image import Image
+
+from pipeline.parameters import Parameters
+
+from improc.bitmask_tools import dilate_bitmask
+from improc.inpainting import Inpainter
+from improc.tools import sigma_clipping
+
+
+class ParsCoadd(Parameters):
+    def __init__(self, **kwargs):
+        super().__init__()
+
+        self.method = self.add_par(
+            'method',
+            'zogy',
+            str,
+            'Coaddition method.  Currently only "naive" and "zogy" are supported. ',
+            critical=True
+        )
+
+        self.alignment = self.add_par(
+            'alignment',
+            {},
+            dict,
+            'Alignment parameters. ',
+            critical=True
+        )
+
+        self.inpainting = self.add_par(
+            'inpainting',
+            {},
+            dict,
+            'Inpainting parameters. ',
+            critical=True
+        )
+
+        self.noise_estimator = self.add_par(
+            'noise_estimator',
+            'sep',
+            str,
+            'Method to estimate noise (sigma) in the image.  '
+            'Use "sep" or "sigma" for sigma clipping. ',
+            critical=True,
+        )
+
+        self.flag_fwhm_factor = self.add_par(
+            'flag_fwhm_factor',
+            1.0,
+            float,
+            'Multiplicative factor for the PSF FWHM (in pixels) to use for dilating the flag maps. ',
+            critical=True,
+        )
+
+        self.enforce_no_new_attrs = True
+        self.override( kwargs )
+
+    def get_process_name(self):
+        return 'coaddition'
+
+
+class Coadder:
+    """Use this class to coadd (stack) images together to make a deeper image.
+
+    Each image should have a PSF and a ZeroPoint associated with it (and loaded!) when running coaddition.
+
+    Images are expected to be aligned (use the Aligner) and should generally be of the same sky region.
+    If not already aligned, they need to have the SourceList and WorldCoordinates loaded so that
+    alignment can be done on the fly.
+
+    Input images should also have a valid Provenance associated with them. Not that for any set of images
+    that share a provenance hash, their respective downstreams (e.g., PSF) should also have a single provenance
+    hash for each type of downstream.  This makes it possible to identify the upstream images' associated products
+    based solely on the provenance's upstream hashes.
+
+    Areas on the edges where the images are not overlapping (or bad pixels, etc.) are coadded but will
+    contribute zero weight, so the total weight of that pixel will be zero (if all input images have a bad pixel)
+    or they would have lower weight if only some images had bad pixels there.
+
+    Remember that coaddition uses convolution with the PSF, so any effects of individual pixels could affect nearby
+    pixels, depending on the size of the PSF.
```

Comment:
...this may be true for the zogy coaddition method, but won't be true for all of them.  (For isntance, the naive coaddition method doesn't do this.)